text/3557-out-of-tree-test-suite.md

```diff
@@ -0,0 +1,359 @@
+
+Out-of-tree testing solves this problem by allowing build and packaging to be performed on one host platform (ex: Ubuntu) while installation and testing are performed on another, compatible platform (ex: Red Hat).
+
+## Certification
+[certification]: #certification
+
+Out-of-tree testing will be essential in enabling tool qualification for certifiable software development in Rust.
+
+Tool qualification involves the demonstration of proper tool operation on a host and a target, where both environments are strictly defined. For DO-178C (avionics safety standard) and ISO-26262 (automotive safety standard), this demonstration involves rigorous and reproducible testing. In particular, evidence must be provided that the version of the tool tested is the version of the tool employed in the certification context.
+
+Any Rust qualification kit would require out-of-tree testing, since out-of-tree testing enables strong assurance that the version of the toolchain delivered is the version of the toolchain tested.
+
+## Authoritative test suite
+[authoritative-test-suite]: #authoritative-test-suite
+
+Considering the ongoing work on the Rust Language Specification, an authoritative test suite will be required to gauge the conformance of a tool against the Rust language. This test suite may be derived from existing Rust tests or written from scratch.
+
+Either way, the authoritative test suite should be executable out-of-tree.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+To address the problems we identified, we propose the use of an out-of-tree test suite (OOTTS). The OOTTS would handle all testing of the Rust toolchain, and replace the existing testing-related Rust infrastructure.
+
+The OOTTS would reside in its own dedicated `rust-lang` repository, and would be considered a distinct Rust product, with its own CI and release artifacts. The OOTTS would consist of infrastructure, test driver, sub-suites, and tests. The infrastructure would handle the building and packaging of the OOTTS. The test driver would be configured via a file, and run requested sub-suites against "testables" (either tools or libraries). Existing sub-suites and tests would be moved out from the existing `rust-lang/rust` repository into the OOTTS repository, and possibly be subjected to light triage.
+
+Users and automation would install the OOTTS and invoke its test driver to verify tools and libraries of their choosing.
+
+Rust developers would use the OOTTS in-tree as a git submodule of `rust-lang/rust`. The existing Rust infrastructure would delegate the building, testing, and packaging of the OOTTS to the OOTTS infrastructure and test driver.
```

Comment:
It needs to be a subtree, not a submodule, so that people can add modify in their Rust PRs.